Better support for documenting expressions with macros

yuyichao opened the pull request:
Fix #11993 (not sure if this is the best way though)
- Update:
  
  See below for the current behavior https://github.com/JuliaLang/julia/pull/12000#issuecomment-118404697 and examples of it https://github.com/JuliaLang/julia/pull/12000#issuecomment-118410489
  
  It still uses `macroexpand` in a `macro`.

Original post:
1. Disallow documenting macro with
   
   ```
   "<doc>"
   @macro
   ```
   
   syntax since it can be ambiguous with expanding and using the macro (which might have side effects)
2. Expand the macro if the expression to be documented is a macro call in order to document the correct object.

Issues
1. This is calling `macroexpand` in a macro. Not sure if this is the best thing to do but I couldn't come up with a better way to do what I want.
2. ~~Both `"" @m` and~~ `""<\n>@m` ~~are~~ is parsed as
   
   ``` julia
   quote
       ""
       @id
   end
   ```
   
   ~~The former should probably be reallowed,~~(edit: fixed) ~~the latter~~ this should probably not have the `block` (although I couldn't figure out if I can reset the input stream)

yuyichao commented:
Updated and `"" @m` is parsed as doc string now. The extra block in `""<\n>@m` is hopefully not a huge issue although I'm still not sure what to do with `macroexpand`.

yuyichao commented:
I could also just document the object if the argument is a macro call but it will fail if this is a macro definition (which probably doesn't matter) and it won't be able to track function signature

```julia
julia> """
       (Float64)
       """
       @inline f(::Float64) = 1
f (generic function with 1 method)

julia> """
       (Int)
       """
       @inline f(::Int) = 1
f (generic function with 2 methods)

help?> f(1)
f(::Int64) at none:4

help?> f
search: f fd for fma fld fft full find filt fill fft! fdio frexp foldr foldl flush

  (Int)
```

JeffBezanson commented:
I'm uncomfortable with making zero-arg macro calls a special case like this. It's very fiddly to need to switch to `"" @m` syntax instead of a newline in this case.

yuyichao commented:
> making zero-arg macro calls a special case like this

So probably
1. `@doc "<doc string>" @macro` should be kept.
2. `"<doc string>" <space or new line> @macro (<with or without arguments>)` should expand the macro and apply the doc macro on the expansion?

JeffBezanson commented:
That sounds better.

yuyichao commented:
There's still the inconsistency that `"<doc string>"<new line or space>a` will be documenting `a` while `"<doc string>"<new line or space>@a` will not document `@a`.

This is the inconsistency I'm trying to address with the 0-argument special case but I guess since it cannot be made fully consistent with everything anyway, I might as well make the implementation simpler....

yuyichao commented:
Oh, and another issue is that I cannot distinguish `@doc "" @macro` and `"" @macro` in the `@doc` macro.

Now I remember why exactly I implement it this way...

1. We want to make `@doc "" @macro` documenting `@macro` (or not?)

2. We want to make `""<\n>@macro<with arguments>` documenting the expansion of `@macro`

This itself will require a special case for 0-argument macrocall.

Furthermore, if there's no special case for this in the parser, `"" @macro` and `""<\n>@macro` will both be lowered to `@doc "" @macro` and will be documenting the `@macro` itself instead of the expansion.

Another way out of this is to make bare string literal document to be lowered to a different internal macro that has different handling of macrocall. This also has the advantage that we and independently tweak the behavior of each one.

yuyichao commented:
In short, in order to get the behavior I mentioned above https://github.com/JuliaLang/julia/pull/12000#issuecomment-118404697 (which I agree is probably the most intuitive one) `"<doc string>" @macro` cannot be lowered to `@doc "<doc string>" @macro`. I'll try to lower it to a different macro and tweak the behavior to see how it looks.

yuyichao commented:
The behavior should be the same with the I described above now.

Here's a compilation of the current behavior

```julia
julia> macro id(f)
       esc(f)
       end

julia> macro m(args...)
       println(args)
       end

# Documenting @id
julia> @doc """
       identity
       """ @id
(anonymous function)

# This is a normal macro call so having a new line does not work
julia> @doc """
       print
       """
       @m
()

help?> @id
  identity

# @m is not documented yet as can be seen from the printing of from @m and that the error below
help?> @m
Symbol not found. Falling back on apropos search ...
No help information found.

# Still won't work because this document the expansion of @m, which is `nothing`
julia> """
       print
       """ @m
()

# As shown here
help?> @m
Symbol not found. Falling back on apropos search ...
No help information found.

# This is the only way to document a macro after it's definition (AFAICT)
julia> @doc """
       print
       """ @m
(anonymous function)

# And it works
help?> @m
  print

# Documenting function definition "decorated" with macros
julia> """
       (Int)
       """
       @inline f(::Int) = 1
f (generic function with 1 method)

julia> """
       (Float64)
       """
       @inline f(::Float64) = 1
f (generic function with 2 methods)

# And the methods are documented correctly
help?> f(1)
  (Int)

# Also for the whole function
help?> f
search: f fd for fma fld fft full find filt fill fft! fdio frexp foldr foldl flush floor float first findn filt! fill! fetch FFTW flipud fliplr fldmod findnz findin

  (Int)

  (Float64)

# Works as expected
julia> """
       """
       @id immutable bar
       x::Int
       end

julia> bar(1)
bar(1)

# Empty doc string. Ooops.
help?> bar
search: bar baremodule SubArray GlobalRef clipboard BitArray backtrace find_library mmap_bitarray BitMatrix takebuf_array catch_backtrace AbstractRNG AbstractArray


# Try again with a different type, now with non-empty doc string.
julia> """
       bar
       """
       @id immutable baz
       x::Int
       end

# And it works.
help?> baz
search: baz

  bar
```

mauro3 commented:
My use case would be to document types defined in macros:
```
julia> macro mkT(f)
       quote
       immutable $f end
       g() = 5 # just a bunch of other stuff also in the macro after the type declaration
       end
       end

julia> @mkT T
#123#g (generic function with 1 method)
```
so this should document the new type:
```
@doc """
My T
"""
@mkT T
```
would this work with this PR? (Sorry for being lazy and not just trying it out!)

yuyichao commented:
> (Sorry for being lazy and not just trying it out!)

NP. I don't always like to wait for ~10min to compile someone else's PR either.... :P 

> would this work with this PR? 

Not directly and not on current master either because

1. A normal macro call cannot use newline to separate arguments.
2. The return value is not the type.

This works.
```julia
julia> macro mkT(f)
       quote
       immutable $f end
       g() = 5
       $f
       end
       end

julia> @doc """
       My T1
       """ @mkT T1
T1

julia> """
       My T2
       """
       @mkT T2
T2

help?> T1
search: T1 Int16 int16 Int128 rot180 int128 UInt16 uint16 Uint16 UInt128 uint128

  My T1

help?> T2
search: T2 hist2d hist2d! Int32 int32 Int128 int128 UInt32 uint32 Uint32 UInt128

  My T2
```

The expansion of the macro is not a single type definition so `@doc` or `@s_doc` will not recognize the structure. However, as long as the expression evaluates to the type, it can document the object with no problem.

P.S. The only exception to the above rules are macro definition (which doesn't return anything) and function definition (which returns the function and not the method, a behavior that is useful in many cases). In these cases, you would have to let `@doc` see the original syntax in order to document the correct stuff.

StefanKarpinski commented:
If the macro expander function were exposed in a more first-class way, then that would be a natural thing to attach the docstring to (in fact, I'm pretty sure this is how it actually works under the hood).

yuyichao commented:
@StefanKarpinski Sorry I think I don't quite understand what you mean here.

What is "a more first-class way" (or why it is not first class now) and this is how what works under the hood?

MichaelHatherly commented:
> I'm uncomfortable with making zero-arg macro calls a special case like this.

The zero-arg case was from when we couldn't add docstrings to most places in base prior to @one-more-minute's ``Base.DocBootstrap`` I think. Now that we can just write

```jl
"""
...
"""
macro m()
    # ...
end
```

anywhere wouldn't it be better to just remove the zero-arg case and require macros to be documented at the definition site? I realise that ``helpdb.jl`` is using this syntax for macros, but hopefully we'll move away from that relatively soon?

That would remove the ambiguity and leave ``:macrocall`` syntax solely for cases such as ``@inline f(x) = ...`` etc.

yuyichao commented:
So what about this (haven't pushed, RFC),

* No special case for macros with or without arguments
* For most (all of) cases, macros should be documented where they are defined.
* If for some reason, it is necessary to document macros after thay are defined, you should pass the symbol directly to `@doc`. This can be done with `@eval` as shown below. It's a bit complicated but should be really rare and doesn't need any special support

I'm using `@eval` for the two cases in `Base` mainly to show that this works and to make the change minimum. If everyone is fine with it, I can move them to the definition of the macro.

```diff
diff --git a/base/docs/Docs.jl b/base/docs/Docs.jl
index 11eade7..df22a5a 100644
--- a/base/docs/Docs.jl
+++ b/base/docs/Docs.jl
@@ -283,21 +283,25 @@ end
 fexpr(ex) = isexpr(ex, :function, :(=)) && isexpr(ex.args[1], :call)
 
 function docm(meta, def)
+    if isexpr(def, :macrocall)
+        def = macroexpand(def)
+    end
     def′ = unblock(def)
     isexpr(def′, :macro) && return namedoc(meta, def, symbol("@", namify(def′)))
     isexpr(def′, :type, :bitstype) && return typedoc(meta, def, namify(def′.args[2]))
     isexpr(def′, :abstract) && return namedoc(meta, def, namify(def′))
     isexpr(def′, :module) && return namedoc(meta, def, def′.args[2])
     fexpr(def′) && return funcdoc(meta, def)
-    isexpr(def′, :macrocall) && (def = namify(def′))
     return objdoc(meta, def)
 end
 
 function docm(ex)
+    if isexpr(ex, :macrocall)
+        ex = macroexpand(ex)
+    end
     isa(ex,Symbol) && haskey(keywords, ex) && return keywords[ex]
     isexpr(ex, :->) && return docm(ex.args...)
     isexpr(ex, :call) && return :(doc($(esc(ex.args[1])), @which $(esc(ex))))
-    isexpr(ex, :macrocall) && (ex = namify(ex))
     :(doc($(esc(ex))))
 end
 
@@ -326,7 +330,7 @@ Please see docs for the `@doc` macro for more info.
 """
 Docs
 
-"""
+@eval @doc """
 # Documentation
 
 Functions, methods and types can be documented by placing a string before the
@@ -363,8 +367,7 @@ Placing documentation before a method definition (e.g. `function foo()
 documented, as opposed to the whole function. Method docs are
 concatenated together in the order they were defined to provide docs
 for the function.
-"""
-@doc
+""" $(symbol("@doc"))
 
 "`doc(obj)`: Get the doc metadata for `obj`."
 doc
diff --git a/base/docs/basedocs.jl b/base/docs/basedocs.jl
index db6a8a4..7b24cc6 100644
--- a/base/docs/basedocs.jl
+++ b/base/docs/basedocs.jl
@@ -286,7 +286,7 @@ keywords[:immutable] = doc"""
   See `type` and the manual for more information.
   """
 
-"""
+@eval @doc """
 Executes an expression, printing the time it took to
 execute and the total number of bytes its execution caused to be
 allocated. Returns the value of the expression. For example:
@@ -295,8 +295,7 @@ allocated. Returns the value of the expression. For example:
       sleep(1)
       2+2
     end
-"""
-@time
+""" $(symbol("@time"))
 
 doc"""
 Construct a regex, such as `r"^[a-z]*$"`. The regex also accepts
```

MichaelHatherly commented:
> It's a bit complicated but should be really rare and doesn't need any special support

+1

yuyichao committed:
Better support for documenting expressions with macros. Fix #11993

1. Always do macro expansion in @doc
2. Use `@doc "<doc string>" :@macro` to document macros after definition
3. Unblock recursively

yuyichao commented:
> Only the `docm(meta, def)` needs macro expansion I think.

Fixed. Thanks.

Also fixed the tests...

mauro3 commented:
A bit slow to respond.  So, if I wanted to document a method defined within a macro I would need to do:
```
macro m(f)
quote
$f(x) = 1
methods(f, (Any,))[1] # potentially tricky if there are many methods
quote
end

@doc """
blah
"""
@m f
```
Right?

yuyichao commented:
@mauro3 I guess you meant two `end` after the `methods` line?

There still canpt be a new line after the doc string if you use the `@doc` macro explicitly (and it's still just now macro are parsed).

I don't think there's currently a way to document a method after it's defined (although I might be wrong). It's probably possible to add support for that (and we should) but that's not in the scope of this PR.

yuyichao commented:
> potentially tricky if there are many methods

Well, if you defined multiple methods (or multiple objects to be documented) in a single macro, there isn't anything the doc system can help with that (unless you call `@doc` in your macro)

If you are just worrying about multiple methods of a single function, it shouldn't be too bad since `which` (and `@which`) should return the `invoke` compatible method now which should be the one that matches the signature precisely and you should be able to generate the correct signatrue in the macro.

mauro3 commented:
Thanks for the clarification.  Works for me.

yuyichao commented:
The document for the `@doc` seems to be on the wrong one. Fixed in this PR.

yuyichao commented:
Another way to make documenting macro after definition less ugly is to turn a string into a symbol in `@doc`, i.e. to document `@macro` with `@doc "<...>" "@macro"` ...

StefanKarpinski commented:
To clarify what I meant – if there was a slicker way to get at the expander function of a macro than

```jl
eval(symbol("@foo"))
```

then that could be what you write after the docstring to document a macro instead of making zero-arg macro after a docstring a special thing.

yuyichao commented:
Well. I think we can add support for documenting `"@foo"` or `:@foo`

yuyichao commented:
Updated:

1. Allow using strings to specify the symbol to be documented. See the corresponding changes in the tests. Hopefully the syntax can make @StefanKarpinski happier.

2. Unconditionally expand the macro in `@doc`. Mainly because the macro might be in a block (to be `unblock`ed) and the expansion of the macro might need `unblock`ing as well. Also, since `macroexpand` works recursively, it is rather inconsistent if we do macro expansion or not depending on whether the toplevel expression is a `macrocall`.

one-more-minute commented:
Expanding macros seems like a reasonable thing to do, but doccing macros as strings isn't going to work. How about having
```julia
macro expander (m)
  @assert isexpr(m, :macrocall) && length(m.args) == 1
  esc(:(eval($(Expr(:quote, m.args[1])))))
end

"foo bar"
@expander @time
```
? Not sure about the name but it should work.

yuyichao commented:
I've thought about that but I didn't go that way because it adds one export and I couldn't figure out where to put it (OK none of them are too big of a deal).

Why wouldn't a string works?

one-more-minute commented:
You can only document a given string once, globally, so every macro doc would have to be qualified – "Base.@time" etc., and when retrieving docs you have to stringify the macro in exactly the right way to check if the string has docs. It's all probably doable in principle but it seems brittle to me, both from an internal and a user's perspective.

yuyichao commented:
I don't think that's the case because `@doc "<...>" "@macro"` is exactly the same with your `@doc "<...>" @expander @macro`. See https://github.com/JuliaLang/julia/commit/672b14a4ac81e0c90716061be6935e05d9b306ac#diff-5a9ad42cbc2e070b47b9d63118501e2fR294

one-more-minute commented:
Ah right, sorry, I didn't see that in the diff. Still, something about notating macros with strings just feels a bit off to me. It's not very general to start with. I don't mind if there's a general consensus that this syntax is the right way forward, though.

one-more-minute commented:
I wonder if it would be better to special case the syntax:
```julia
"foo"
:@time
```

MichaelHatherly commented:
How about quoting it?

```jl
"..."
:(Base.Libc.@foo)
```

yuyichao commented:
So by "I couldn't figure out where to put it" I really mean this (`@expander`) feels like something generic but I couldn't figure out what else is it useful for and what behavior should it really have if it's not something internal to the doc system anymore. Just a macrocall remover?

one-more-minute commented:
I think `Base.Meta` would make sense, or you could put it in `Base.Docs` and have people explicitly import it when they need it (since it wouldn't be that common). I like the quoted syntax a lot though, and it looks like @MichaelHatherly's thinking on the same wavelength, so that seems like the best option so far.

yuyichao commented:
That `:@macro` was also one of my option. Just feel like it's less intuitive for people who don't know much about quoting and it's harder to convince myself that it is not ambiguous with anything else. (since there are a lot of syntax / situations for quoting)

yuyichao commented:
In another word, I have thought about special case `@doc "<...>" :@macro` but then

1. Should this syntax be supported for retrieving the doc?
2. Should this be supported for symbols as well (i.e. `@doc "<...>" :other_object`)
3. Should this be supported for quote note in general?

I think we can do it. Just need more thought. (And I don't mind going through these now)

one-more-minute commented:
In terms of retrieval I think we should carry on special casing the zero-arg form. It may not be perfectly intuitive but we can't make newbies type `?:@time` or `?"@time"` in the repl.

Actually, there's a solid argument against strings – autocompletion.

As for storage I think we should look for the exact forms `Expr(:quote, Expr(:macrocall, _))` and `Expr(:quote, Expr(:macrocall, _, ""))`, and ignore everything else. (We may throw an error when documenting values / unrecognised syntax eventually if it causes confusion.)

MichaelHatherly commented:
> less intuitive for people who don't know much about quoting

If someone is documenting a macro then I guess they probably would have also written it, in which case I'd think they are probably aware of quoting. But yes, it's not totally obvious immediately.

> In terms of retrieval I think we should carry on special casing the zero-arg form. It may not be perfectly intuitive but we can't make newbies type ``?:@time`` or ``?"@time"`` in the repl.

Should ``@doc`` be used for both registering *and* retrieving docs? It could possibly make things a little clearer if we split them into two different macros.

yuyichao commented:
> In terms of retrieval I think we should carry on special casing the zero-arg form. It may not be perfectly intuitive but we can't make newbies type ?:@time or ?"@time" in the repl.

I'm not talking about removing that. I was just wondering if `@doc :@time` should also be supported in additional to `@doc @time`

> Actually, there's a solid argument against strings – autocompletion.

Good point. Hopefully no one will have a super long macroname that need autocompletion to document though .....

> As for storage I think we should look for the exact forms Expr(:quote, Expr(:macrocall, _)) and Expr(:quote, Expr(:macrocall, _, "")), and ignore everything else. (We may throw an error when documenting values / unrecognised syntax eventually if it causes confusion.)

Fine with me.

yuyichao commented:
> Should @doc be used for both registering and retrieving docs? It could possibly make things a little clearer if we split them into two different macros.

There's no ambiguity for now so it's probably fine to keep it like this and not introducing any breakage.

one-more-minute commented:
Sure, I'm definitely on board with supporting `@doc :@time`.

@MichaelHatherly At the moment it's reasonable to think of `@doc x` and `@doc y x` as being separate macros – renaming them would just be changing the label. The original reasoning was that it's more intuitive and concise – one less name to remember when you want to get to docs quickly – but now that the `@doc y x` form is hidden behind syntax sugar anyway there's a reasonable case for renaming it.

MichaelHatherly commented:
> The original reasoning was that it's more intuitive and concise – one less name to remember when you want to get to docs quickly – but now that the ``@doc y x`` form is hidden behind syntax sugar anyway there's a reasonable case for renaming it.

Sure, makes sense. I'll investigate whether changing it simplifies the implementation when I've got some time.

yuyichao commented:
Back from meeting...

The `macroexpand` doesn't play well with `:@doc`

```julia
```

yuyichao commented:
> Sure, makes sense. I'll investigate whether changing it simplifies the implementation when I've got some time.

I'd prefer not breaking it if it is not necessary. I'm fine either way though.

one-more-minute commented:
That's pretty weird. How about `unblock`ing once before macro expansion, checking for the macro special case, then falling through to what we already have?

yuyichao commented:
> That's pretty weird. How about unblocking once before macro expansion, checking for the macro special case, then falling through to what we already have?

Something like that, yes. I'm just listing the cases we want to support and see how I can handle all of them.

I'm also wondering if `unblock` should be recursive. Especially for the result of a macro expansion.

one-more-minute commented:
Yes, I was just thinking that myself. It should be fine for `unblock` to call itself just before returning, if it's returning a new expression.

yuyichao commented:
I guess if someone write sth recursive like `ex = Expr(:block); push!(ex.args, ex)` we have bigger problems =)

one-more-minute commented:
Haha, how have I never tried building a recursive expression before... there must be some crazy use for this (if nothing else the output in the repl is kinda cool to watch)

yuyichao commented:
> there must be some crazy use for this

One of the few ways to drive the parser crazy?

Anyway, I listed the procedure of the expansion https://github.com/JuliaLang/julia/commit/edd46a37ccd6d37e7ce87aedf792e10fc5b38a02#diff-5a9ad42cbc2e070b47b9d63118501e2fR287 and write  what I think is a pretty minimum implementation of it...

I'm [allowing more general quoted macrocall for now](https://github.com/JuliaLang/julia/commit/edd46a37ccd6d37e7ce87aedf792e10fc5b38a02#diff-5a9ad42cbc2e070b47b9d63118501e2fR297) mainly because I'm too lazy to check for all cases. We can disallow this if there's confusion. (I really doubt...)

Tests pass locally. Let's how the CI think...

yuyichao commented:
Just realized that I can do manual TCO for `unblock`....... LOL
Will do that if this needs another revision (likely?).

one-more-minute commented on base/docs/Docs.jl:
```diff
@@ -201,7 +201,7 @@ end
 doc(f, ::Method) = doc(f)
 
-# Modules
+# Modules
```
This is pretty weird, have we somehow ended up with a different kind of space here?

yuyichao replied:
It's a non-breaking space. We have a lot of them.

I have an old PR to clean them up but I still need to rebase it and check if any of them are used intentionally.

yuyichao replied:
Ref https://github.com/JuliaLang/julia/pull/11605

one-more-minute replied:
I assume that must be my fault, although I have no idea why an editor would insert one of those.

yuyichao replied:
Editor shortcut probably?

one-more-minute replied:
Ah yes, looks like Alt-Space does it on OS X – better watch out for that one.

one-more-minute commented:
I imagine there are many people who would want to see a manual implementation of TCO, but I'm definitely not one of them :) As a bonus we throw an error when people try to documenting infinite expressions, rather than hanging.

Anyway, LGTM, I'll wait on the tests.

yuyichao commented:
@tkelman I guess the push failure is sth know and is fixed now?

tkelman commented:
yeah, fixed by 896cadf4555efcc42478780bf30fcc1376d879fa

yuyichao commented:
Since this seems to be getting ready for merging, I created a `@compat` entry https://github.com/JuliaLang/Compat.jl/pull/113 in case any packages is using the old syntax to document macros.

mauro3 commented:
Would this need documenting too?

yuyichao commented:
I didn't bother add doc yet mainly because this wasn't a documented interface to begin with and originally this change distinguish between documenting with or without `@doc` macro and the new interface isn't documented yet either =(.

This isn't the case anymore so if everyone is happy about the current state I'll add some document about the change soon.

yuyichao committed:
Documenting the new macro documentation interface

yuyichao commented:
Added a section about documenting after an object is defined in the documentation of `@doc`. I'm pretty bad at writing documentation so suggestions welcome.

one-more-minute commented:
I'm wondering if it'd be a good idea to put [ExpressionMatch.jl](https://github.com/one-more-minute/ExpressionMatch.jl) into Base. Especially if we start adding more syntax to `@doc` (and being more finicky about it), it's only going to get more obscure, and we could make it a lot clearer what's going on:
```julia
julia> @match :(:@time) begin
         :(@f_) -> "macro $f"
         :(f_"") -> "string macro $f"
       end

"macro @time"

julia> @match :(:(r"")) begin
         :(@f_) -> "macro $f"
         :(f_"") -> "string macro $f"
       end
"string macro @r_str"
```

yuyichao commented:
> I'm wondering if it'd be a good idea to put `ExpressionMatch.jl` into Base.

I'm not an expert on what should go into base. But I'd say not this PR for sure.

> Especially if we start adding more syntax to `@doc` (and being more finicky about it).

Well, OTOH, making it harder to add syntax also makes it harder to introduce breaking changes and obscure syntax. It took me a long time to convince myself that it is ok to add new syntax like this and I hope this should happen as infrequent as possible.

one-more-minute commented:
Sure, definitely not for this PR.

I just introduced a merge conflict (sorry), but if you can rebase this I'll merge then.

yuyichao commented:
> I just introduced a merge conflict (sorry), but if you can rebase this I'll merge then.

I can tell you merge that PR just to post this comment :laughing: 

It's a stupid end-of-test-file conflict. I'll push after it passes local tests.

one-more-minute commented:
Nothing interesting has changed since this passed the tests before, right? I'd really like to start working off these changes.

yuyichao commented:
@one-more-minute There shouldn't be.

The AppVeyor test should be done in a few minutes so I would suggest waiting for a few more minutes rather than being ping'd by Tony if sth unexpected happened.

one-more-minute commented:
Awesome, really glad we have a better approach for this now.

yuyichao commented:
@IainNZ Could you please check if this causes any breakage in packages? Depending on how bad it is we can either merge or discard JuliaLang/Compat.jl#113
